File: packages/python/src/kenya_data/counties.py
"""County lookups."""
from __future__ import annotations

import builtins
import sqlite3

from .exceptions import EntityNotFoundError
from .models import County


class Counties:
    def __init__(self, conn: sqlite3.Connection) -> None:
        self._conn = conn

    def list(self) -> builtins.list[County]:
        rows = self._conn.execute(
            "SELECT id, code, name, capital, status FROM counties ORDER BY name"
        ).fetchall()
        return [County(**dict(r)) for r in rows]

    def get(self, name_or_code: str) -> County:
        row = self._conn.execute(
            "SELECT id, code, name, capital, status FROM counties "
            "WHERE name = ? COLLATE NOCASE OR code = ? COLLATE NOCASE",
            (name_or_code, name_or_code),
        ).fetchone()
        if row is None:
            raise EntityNotFoundError("county", name_or_code)
        return County(**dict(row))

File: packages/typescript/src/counties.ts
import type Database from "better-sqlite3";
import { EntityNotFoundError } from "./errors.js";
import type { County } from "./types.js";

interface CountyRow {
  id: number;
  code: string;
  name: string;
  capital: string | null;
  status: string;
}

function toCounty(row: CountyRow): County {
  return { id: row.id, code: row.code, name: row.name, capital: row.capital, status: row.status };
}

export class Counties {
  constructor(private readonly db: Database.Database) {}

  list(): County[] {
    const rows = this.db
      .prepare("SELECT id, code, name, capital, status FROM counties ORDER BY name")
      .all() as CountyRow[];
    return rows.map(toCounty);
  }

  get(nameOrCode: string): County {
    const row = this.db
      .prepare(
        "SELECT id, code, name, capital, status FROM counties " +
          "WHERE name = ? COLLATE NOCASE OR code = ? COLLATE NOCASE",
      )
      .get(nameOrCode, nameOrCode) as CountyRow | undefined;
    if (!row) throw new EntityNotFoundError("county", nameOrCode);
    return toCounty(row);
  }
}
